### PR TITLE
Fix confusing HDF5_OPTIONAL

### LINKED ISSUE #17: Desired behavior of HDF5_OPTIONAL
In https://github.com/FrescolinoGroup/pyhdf5io/commit/876541318273ec74f239ffb8fcf23ec22cef5e9a I introduced (without PR, heh..) a class attribute `HDF5_OPTIONAL` to the `SimpleHDF5Mapping`. When a key is in `HDF5_OPTIONAL`, it would be skipped in saving / loading if not present.

**BUT:** in order to be serialized in the first place, the key **also** needs to be in `HDF5_ATTRIBUTES`. I now think this is a bit unfortunate, because it's really easy to get wrong. Case in point: of two uses I found in my codes, one is incorrect.

I would suggest making the two lists *mutually exclusive* (and checking for that) -- this is a backwards-incompatible change, but worth it IMO.

Pinging @mskoenz for opinion.

### PR DESCRIPTION
**Pull request by greschd**
Fixes #17.

This change makes keys in `HDF5_ATTRIBUTES` and `HDF5_OPTIONAL` mutually exclusive, since that is the more reasonable option. The check happens only when saving or loading, because otherwise it would need to be implemented in a metaclass - or the check would happen in a class decorator, which is kind of useless (the class decorator can be forgotten just as easily).

Since the change is backwards incompatible, bump the version number to `1.0.0`.

### CODE REVIEW

**Comment by greschd** [comment]
Will make an update to https://github.com/greschd/NodeFinder necessary, but fixes incorrect usage in https://github.com/Z2PackDev/symmetry_representation.

File: fsc/hdf5_io/_special_types.py [hunk 1]
```diff
@@ -185,4 +185,4 @@ def _to_hashable(obj):
 
 @_to_hashable.register(Iterable)
 def _(obj):
-    return tuple(obj)
+    return tuple([_ensure_hashable(val) for val in obj])
```
Comment (mskoenz) [inline fsc/hdf5_io/_special_types.py:188]: What is this change, how is it related to the optional attribute change?

File: tests/test_save_load.py [hunk 2]
```diff
@@ -2,6 +2,7 @@
 Tests for saving and loading a simple class.
 """
 
+import os
 import tempfile
```
Comment (mskoenz) [inline tests/test_save_load.py]: why not [pathlib](https://docs.python.org/3.7/library/pathlib.html?highlight=path#module-pathlib)?
``` python
from pathlib import Path
path = Path("folder", "file")
```

Comment (mskoenz) [reply tests/test_save_load.py]: (I know its just tests, but it is imo much nicer)

File: fsc/hdf5_io/_special_types.py (hunk 1 above)
Comment (greschd) [reply fsc/hdf5_io/_special_types.py:188]: It's not directly related, but for the key to be hashable contents of e.g. a tuple need to be hashable also.

File: tests/test_save_load.py (hunk 2 above)
Comment (greschd) [reply tests/test_save_load.py]: Yeah, I could just change the `sample` fixture to return a `pathlib.Path` (maybe rename it to `sample_dir`), and then do

`sample_dir / 'folder' / 'file'`

Comment (greschd) [reply tests/test_save_load.py]: Done.

**Comment by greschd** [comment]
Aah py3.5, bane of my existence 😄

**Comment by mskoenz** [comment]
Its failing in 3.5 bc open(Path(), ...) does not work. It needs to be a string (which os.path.joinpath provides I guess). Needs to be str(Path(...)) its super annoying

**Comment by mskoenz** [comment]
Drop 3.5 support?

File: fsc/hdf5_io/_special_types.py (hunk 1 above)
Comment (mskoenz) [reply fsc/hdf5_io/_special_types.py:188]: ok

File: fsc/hdf5_io/_version.py [hunk 3]
```diff
@@ -2,10 +2,7 @@
...
 """
 
-import os
+import pathlib
 
-with open(
-    os.path.join(os.path.dirname(os.path.abspath(__file__)), 'version.txt'),
-    'r'
-) as f:
+with open(pathlib.Path(__file__).parent.resolve() / 'version.txt', 'r') as f:
```
Comment (mskoenz) [inline fsc/hdf5_io/_version.py]: should work, but .resolve().parent is safer

Comment (mskoenz) [reply fsc/hdf5_io/_version.py]: nvm, always works

Comment (greschd) [reply fsc/hdf5_io/_version.py]: How so?

**Review by mskoenz — DISMISSED** [review]
All good, nicer now with pathlib

File: fsc/hdf5_io/_version.py (hunk 3 above)
Comment (mskoenz) [reply fsc/hdf5_io/_version.py]: its a file, which always has a parent

Comment (mskoenz) [reply fsc/hdf5_io/_version.py]: for folders you might run out of parent before resolve, but never tested it, just a hunch

Comment (greschd) [reply fsc/hdf5_io/_version.py]: Ah, recommendation is here: https://docs.python.org/3.5/library/pathlib.html#pathlib.PurePath.parent

Comment (mskoenz) [reply fsc/hdf5_io/_version.py]: Yes, so that was it, but we can leave it here, just for the future

Comment (greschd) [reply fsc/hdf5_io/_version.py]: Well, better to build a good habit.

**Review by mskoenz — APPROVED** [review]
all good

**Comment by greschd** [comment]
Thanks for the review! Agree also on the bump to 1.0 version?

**Comment by mskoenz** [comment]
Np

**Pull request merged**